_posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd

```diff
@@ -0,0 +1,390 @@
+
+```{r}
+jit_save(f_t, "/tmp/f_t")
+
+f_t_new <- jit_load("/tmp/f_t")
+```
+
+## A first glance at optimizations
+
+Optimizations performed by the `torch` JIT compiler happen in stages. On the first pass, we see things like dead code elimination and pre-computation of constants. Take this function:
+
+```{r}
+f <- function(x) {
+  
+  a <- 7
+  b <- 11
+  c <- 2
+  d <- a + b + c
+  e <- a + b + c + 25
+  
+  
+  x + d 
+  
+}
+
+
+```
+
+Here computation of `e` is useless -- it is never used. Consequently, in the intermediate representation, `e` does not even appear. Also, as the values of `a`, `b`, and `c` are known already at compile time, the only constant present in the IR is `d`, their sum.
+
+Nicely, we can verify that for ourselves. To peek at the IR -- the initial IR, to be precise -- we first trace `f`, and then access the traced function's `graph` property:
+
+```{r}
+f_t <- jit_trace(f, torch_tensor(0))
+
+f_t$graph
+```
+
+    graph(%0 : Float(1, strides=[1], requires_grad=0, device=cpu)):
+      %1 : float = prim::Constant[value=20.]()
+      %2 : int = prim::Constant[value=1]()
+      %3 : Float(1, strides=[1], requires_grad=0, device=cpu) = aten::add(%0, %1, %2)
+      return (%3)
+
+And really, the only computation recorded is the one that adds 20 to the passed-in tensor.
+
+So far, we've been talking about the JIT compiler's initial pass. But the process does not stop there. On subsequent passes, optimization expands into the realm of tensor operations.
+
+Take the following function:
+
+```{r}
+f <- function(x) {
+  
+  m1 <- torch_eye(5, device = "cuda")
+  x <- x$mul(m1)
+
+  m2 <- torch_arange(start = 1, end = 25, device = "cuda")$view(c(5,5))
+  x <- x$add(m2)
+  
+  x <- torch_relu(x)
+  
+  x$matmul(m2)
+  
+}
+```
+
+Harmless though this function may look, it incurs quite a bit of scheduling overhead. A separate GPU *kernel* (a C function, to be parallelized over many CUDA threads) is required for each of `torch_mul()` , `torch_add()`, `torch_relu()` , and `torch_matmul()`.
+
+Under certain conditions, several operations can be chained (or *fused*, to use the technical term) into a single one. Here, three of those four methods (namely, all but `torch_matmul()`) operate *point-wise*; that is, they modify each element of a tensor in isolation. In consequence, not only do they lend themselves optimally to parallelization individually, -- the same would be true of a function that were to *compose* ("fuse") them: To compute a composite function "multiply then add then ReLU"
+
+$$
+relu() \ \circ \ (+) \ \circ \ (*)
+$$
+
+on a tensor *element*, nothing needs to be known about other elements in the tensor. The aggregate operation could then be run on the GPU in a single kernel.
+
+To make this happen, you normally would have to write custom CUDA code. Thanks to the JIT compiler, in many cases you don't have to: It will create such a kernel on the fly*.*
```

Comment:
Are you emphasizing the period on this sentence on purpose? If so, I'm not sure I understand why.

Reply:
oops, idk how that happened ...? will remove